[precompile] Add Exception type PackageError for unsupported precompile features.

zhxchen17 opened the pull request:
Summary:
Today when guard serialization fails, dynamo will raise an internal error like:

```
torch._dynamo.exc.InternalTorchDynamoError: RuntimeError: CLOSURE_MATCH guard cannot be serialized.
```

Adding a dedicated PackageError type to surface the error more clearly.

Test Plan: CI

Differential Revision: D75452124




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Skylion007 commented on torch/_dynamo/exc.py:
```diff
@@ -271,6 +271,10 @@ class FailOnRecompileLimitHit(Exception):
     pass
 
 
+class PackageError(Exception):
```
Should this subclass Exception? NotImplementedException? RuntimeException?

jamesjwu replied:
Maybe just TorchDynamoException makes sense here

jamesjwu commented on torch/_dynamo/guards.py:
```diff
@@ -2704,7 +2716,10 @@ def reducer_override(self, obj):
-    pickler.dump(state)
+    try:
+        pickler.dump(state)
+    except AttributeError as e:
```
Is AttributeError the only type of pickling error we need to catch?

zhxchen17 replied:
Yes, although it's weird that pickle throws attribute error when it cannot handle local functions.

zhxchen17 committed:
[precompile] Add Exception type PackageError for unsupported precompile features. (#154430)

Summary:

Today when guard serialization fails, dynamo will raise an internal error like:

```
torch._dynamo.exc.InternalTorchDynamoError: RuntimeError: CLOSURE_MATCH guard cannot be serialized.
```

Adding a dedicated PackageError type to surface the error more clearly.

Test Plan: CI

Reviewed By: jamesjwu

Differential Revision: D75452124
